### ISSUE TITLE
no time.sleep() ... how to call js async code from python runtime ?

### ISSUE THREAD

**Issue by amigrave** [pyodide/pyodide#97]
I noticed that `time.sleep()` is a no op in the pyodide's cpython webassembly runtime. As cpython implements it using `select.select()`'s timeout, it's not really a surprise that it's not supported by pyodide.

That said I have to call a javascript function from the python runtime and the js function will do asynchronous stuff before being able to return the value. eg:
```python
from js import my_async_javascript_function
```

I first tried to do polling but without `time.sleep()` it's not doable.

How can I handle this properly with pyodide ? Through Promise ? async/await ?  ... I could not find a clue for this, any hint would be appreciated.

Thanks

**Comment by mdboom** [comment]
This is definitely something on the roadmap for which there is currently no good solution.

The best I can recommend is to write a wrapper around the Javascript function (in Javascript) that takes a callback function as an argument, and then calls the callback when the async function is complete.  From Python, you can then pass a Python function in as an argument, and Python will continue running from there.

An alternative (if you really want to do polling) is to use Javascript's `window.setTimeout` from Python, and poll in the callback function.

Both of these options have to delegate control to a callback function, however, which is why neither is a complete solution.

**Comment by amigrave** [comment]
@mdboom thanks for the advise ! Shall I keep the issue opened ?

**Comment by mdboom** [comment]
I think it's worth keeping it open -- we should have a better solution for this, or at least better documentation.

**Comment by daicoden** [comment]
Poking around this....

Looks like upgrading pyodide to work on llvm #476 would enable asyncify's emscripten_sleep.

https://kripken.github.io/blog/wasm/2019/07/16/asyncify.html

Is the work to LLVM important, or nice to have?

**Comment by rth** [comment]
So asyncify apparently has around 50% performance cost which could be a lot for default builds. From,  https://emscripten.org/docs/porting/asyncify.html#optimizing
> Asyncify adds overhead, both code size and slowness, because it instruments code to allow unwinding and rewinding. That overhead is usually not extreme, something like 50% or so. 

but we won't find out how much it would be for pyodide until we try.

In any case, updating to the LLVM upstream emscripten backend it is essential, and should unblock a number of issues. It tuned out to be not easy so far. I'll comment in https://github.com/iodide-project/pyodide/issues/476 with a status update.

**Comment by pmp-p** [comment]
>  That overhead is usually not extreme, something like 50% or so. 

Sadly that is for normal programs where only a few blocking calls need to be async. 

In an interpreter VM : the whole path to - and including - the bytecode loop must be "asyncified" because it can call eg sleep anytime.
With a pre-emptible stackless interpreter, which does not have to record the stack, only instruction pointers, the best loss is 2-3x, while generic asyncify is around 10x.

**Comment by pmp-p** [comment]
@amigrave problem is linked to the fact that time.sleep() in cpython is just an idle wait, it allows OS to pause the Python thread, and process IO meanwhile.

If you are not in a worker thread on wasm you can't do that way : that means exiting cPython loop to give hand to javascript host

Doing that will unwind the stack and using setjmp was no help last time i tried.

an alternative solution for cpython in its current state - without using worker - is to use asyncio for everything that **you** can predict will block in your python code.

**Comment by sconover** [comment]
> If you are not in a worker thread on wasm you can't do that way : that means exiting cPython loop to give hand to javascript host

@pmp-p What would your approach be to implementing sleep via a webworker? Maybe I'm being dense but the only solution that comes to mind is a spinlock, but I assume that such a lock would also prevent communication via `onmessage`/`postMessage`...

**Comment by pmp-p** [comment]
@sconover i'm sorry i have no real clear idea for that : web workers/asyncify are outside of my experimentation field, my far goal is multiple wasm/wasi interpreters loops running as uniprocessor couroutines on single hardware thread each.

 Looking at pyodide from that point of view ( not web centric, and refusing threading ) I would pick up stackless python code base instead of cpython  and make it pre-emptible in ceval.c, so any blocking bytecode block could yield to host via the vm eval loop via aio_suspend instead of python loop.

But that's kinda dreaming so to summarize what is real at the moment :

my wapy python interpreter model has two vm :
 1 - original one with C-stack full speed for asyncio use only. 
 2 - mine is a C-assembly coroutine built upon pfalcon stackless model with as much registers as recursion depth and is 2x slower than async ie 4-7x slower than native cpython running in a thread. 
when you want IO/timers , you launch io request on async loop, call aio_suspend() until your IO is serviced or timeout. ( here is sleep in wapy https://wyz.fr/0L-andin )

 The asyncio loop gives access to timers and io completion various locks, but it is running **interleaved** with the sync loop which gives hand back to js with the aio_suspend() function

 so it's not spinlock + busy looping. but more a python vm loop as a C coroutine which resume instead of yielding depending on locks conditions.

To achieve that on cpython without core changes, I think you could POC implementing a Python VM in the python asyncio loop to run the sync bytecode ( there are some python-in-python vm around but not sure if any of those were written with async in mind )

i think such a POC would be insightfull for people that don't get fully why the actual cpython loop vm model is becoming a real blocker for some domains of application (including but not limited to real time).

**Comment by pmp-p** [comment]
it's proven possible to run sync code with pyodide  without modifying C-Python
poc, not fully compliant  https://github.com/pmp-p/aiovm.

**Comment by joemarshall** [comment]
Rather than run a python VM, it is easy enough to run some really basic scheduling in javascript, and then call into python to run asynchronous code just like asyncio does.

I've hacked together something here - it uses stock pyodide, plus a little bit of javascript, and you can
a) run async / await code with a basic implementation of coroutines, futures etc.
b) run normal synchronous code with sleep, print in, and it translates the code on compile to async code so that sleeps work in the browser.

Have a play here:

https://joemarshall.github.io/pyodide-async/

**Comment by pmp-p** [comment]
> translates the code on compile to async code 

@joemarshall  That is just known bytecode patching no ? could that solve deadline preemption in infinite loops to prevent page getting killed too ?

**Comment by joemarshall** [comment]
No, the make code async thing patches the source tree using ast.

The underlying functionality is just similar to asyncio, in that it supports running multiple coroutines in the main thread, but similar to that, it only returns back to JavaScript each time you do an await on something. If your infinite loop has a sleep in it, then you're grand, and if it has a call to an async function, again you're cool, but otherwise it doesn't drop back unless explicitly allowed. I could wrap while and for loops as async loops, that wouldn't be super hard. Doing that would mean any loop based code would not lock the JavaScript thread. I don't know what the performance hit would be though.

**Comment by pmp-p** [comment]
@joemarshall thanks for the clarifications 

>  I don't know what the performance hit would be though.

it cannot be worse than asyncify (10x). native vm async will always be the fastest way. The closer you can get to the [Control Flow](https://en.wikipedia.org/wiki/Call_graph) the faster you will be, emsdk asyncify was not made for dynamic dispatch.

>  I could wrap while and for loops as async loops, that wouldn't be super hard

That is what running another VM as async loop does automatically, but it does that brute-force for **everything** and if written in pure python cost is very high ( though in C il would be **better** than asyncify). So it is definitevely worth investigating !

Unless rewriting byterun (stuck on 3.7) as a C coroutine, with all the maintenance burden ( no existing maintener stood up for the python version, nedbat already declined  ) : your approach seems a good tradeoff.

It would be really nice to have a cpython poc with dunder removed ( maybe use "aio" namespace because it's finally what it's all about ) and some test runner because i'm afraid there will be a lot of corner cases ( same those motivated async/await to be the way it is  :/ ).

off course none of those "py-asyncify" or "vm-async" solutions will solve the blocking I/O calls in native modules  but a starting point is **better than nothing**. I have wapy vm as fallback solution very well suited for light web but i'm ok to jump on the first train to have cpython too, there can't be too much python until there's enough.

**Comment by joemarshall** [comment]
Wait a minute, I've just realised all the asyncio stuff is still in there, it just doesn't work because it stays stuck in python land. It just needs a small JavaScript hook to run the run once method in the loop. So it didn't need all the reengineering I did!

What probably makes sense is to make it so asyncio works with JavaScript, then leverage that to allow web workers to communicate with main JavaScript thread while python code is running.

**Comment by pmp-p** [comment]
>  It just needs a small JavaScript hook to run the run once method in the loop.

That's how I do on Panda3D webgl demos, i already have asyncio/uasyncio (3.7/3.8/3.9) partially patched for wasm. It would be really nice to have you around https://gitter.im/Wasm-Python/community dunno if you got the invite

**Comment by rth** [comment]
Very interesting work @joemarshall and @pmp-p!

I think it would be great having a documentation page discussing current state of async code in pyodide, linking to your projects and either giving some examples of how to use them or linking to your documentation examples. Would you be interested in contributing it?

**Comment by joemarshall** [comment]
Yeah, sounds like a good idea. I'm part way through a) tidying up my code so that it just uses straight asyncio, which works fine and b) playing with something else - a simple (ish) unthrowable exception - it can be raised, then you can unthrow back to where it came from. The good about that is that it is then pretty straightforward to replace time sleep with an unthrowable which drops out to JavaScript then is unthrown back at the right time, allowing non async code to run nicely with minimal modification . It uses a trace function which means during the unthrow it may be slow, especially if it is unthrowing through with, for or try blocks where I need to use bytecode messing, but given the use case for this of needing to deal with a) delays and b) input callbacks, and the way once the stack rewinds it has zero performance overhead, I think it will be a nice little toy.

**Comment by joemarshall** [comment]
I did some tidying at  https://joemarshall.github.io/pyodide-async/ 

It now runs with asyncio primitives instead of my custom ones.  I also did some timing tests of the asyncify which you can play with on there - it appears just converting calls to async has a bit of overhead. For a blank function call that is asyncified it increases by roughly 80%. However if asyncify something with anything useful in the function body, even something really short like adding and multiplying in a loop, that overhead drops to approx 10%.

At the moment I'm only asyncifying the user submitted code not everything in modules called from it. I think that makes sense from a performance point of view. From a practical point of view it means no async calls in callbacks from non-user code, but you can add a future onto the event loop should you need to do that.

**Comment by rth** [comment]
Resolved in https://github.com/pyodide/pyodide/pull/1158 and https://github.com/pyodide/pyodide/pull/880 (among other PRs).

**Comment by chrispiech** [comment]
Very interested in this thread. @rth is it possible to use time.sleep now?

Thank you all for your hard work.

**Comment by rth** [comment]
@chrispiech This was closed because a lot of the discussion went more about async support which is now resolved. Currently, you can sleep with,
```py
await asyncio.sleep(duration)
```
For `time.sleep` in sync code I opened https://github.com/pyodide/pyodide/issues/2354
